Security

cryogenian opened the pull request:
Currently only stubs and specs are ready for this in slamdata/quasar-advanced#50 and I'm not sure if this should be merged before sharing is implemented in quasar-advanced. Although this won't harm any existed code probably. 

@garyb Could you please take a look?

garyb commented on src/Quasar/Advanced/QuasarAF.purs:
```diff
@@ -19,87 +19,295 @@ module Quasar.Advanced.QuasarAF
+  | TokenInfo Qa.TokenId (Qa.Token :~> a)
+  | CreateToken (Maybe Qa.TokenName) (Array Qa.Action) (Qa.Token :~> a)
+  | DeleteToken Qa.TokenId (Unit :~> a)
+  | AuthProviders ((Array Qa.Provider) :~> a)
```
You probably noticed there was a problem with how tightly the type operators bind, that'll be fixed in 0.9...

cryogenian replied:
Yeah, I've seen that issue. That would be great!

garyb commented on src/Quasar/Advanced/Types.purs:
```diff
@@ -0,0 +1,393 @@
+      -- ¯\_(ツ)_/¯
+      if isJust (Str.indexOf "@" str)
+        then pure str
+        else Left "Incorrect email"
```
Haha, is this validation something we'll need to revisit perhaps then? Can we not assume that when decoding it's already "safe"?

cryogenian replied:
Unfortunately not :( `grantedTo` field is `group|tokenId|userEmail`. It's better to add real email parser of course, but checking `@` should be enough for now I think.

garyb replied:
Ah ok, I see. Perhaps we can add a more descriptive comment along with the shrug then? :laughing:

cryogenian replied:
Sure thing :smile:

garyb commented on src/Quasar/Advanced/Types.purs:
```diff
@@ -0,0 +1,393 @@
+    { tokens ∷ Array TokenId
+    , users ∷ Array UserId
+    , groups ∷ Array (Pt.AbsFile Pt.Sandboxed)
+    }
```
I'd prefer to avoid `newtype`s around records in favour of plain records if we can, to match the other cases where this occurs.

Unless these records are internal to the quasar stuff and not actually useful for consumers of the API, in which case `newtype` away if it makes things easier!

garyb commented on src/Quasar/Error.purs:
```diff
@@ -18,20 +18,30 @@ module Quasar.Error where
 
 printQError ∷ QError → String
 printQError NotFound = "Resource not found"
+printQError Forbidden = "Resource is unavailable, please, authorize"
```
Perhaps "Resource is unavailable, authorization is required first" or something like that? If "please authorize" is presented to a user at some point they might think it is something they can do, whereas the likelihood is it's something we've done wrong.

garyb commented:
Apart from those few comments the code all looks good to me. :+1:

I'd like to have at leas one super simple test case for each of the advanced endpoints running before we merge it though, as we have for the community stuff currently, so I guess we need the new quasar-advanced for that first too.

cryogenian commented:
- Changed message in `Quasar.Error`
- Changed comment for `UserId` decoding 
- API works with records, newtypes used internally for encoding/decoding. 

Unfortunately backend isn't ready yet --> I can't implement tests. Could we push this to separate branch? Or maybe just don't close this pr until tests? Or merge and don't release?

cryogenian commented:
I've left newtypes for special `String` cases like `TokenHash` and `PermissionId` though

garyb commented:
Thanks for the amends! 

String newtypes are :ok_hand:, it's more to avoid the need for lenses to make record manipulations convenient.

garyb commented:
Oops, posted that ^ before I was done typing.

As for tests, I realise that they're not implementable yet, I was just saying ideally before we merge this we'd have something to test against. If we need this to make progress with SlamData maybe we could reference your branch in the `bower.json`?

I'm fine with just keeping the PR open until we have tests - I don't think there's anything else we need from `purescript-quasar` right now, so we shouldn't have any conflicts at least!

cryogenian commented:
Ok, thanks for review!

garyb commented:
I'll go ahead and merge this since we've been using it a while anyway :wink: (and I need to update this for 0.9)

The build error is just a `Debug.Trace` reference, so I'll remove that while updating.
